rest-server/core/src/main/scala/com/actionml/core/jobs/JobManager.scala

```diff
@@ -45,87 +46,78 @@ trait JobManagerInterface {
-    jobDescriptions = jobDescriptions + (engineId -> newJobDescriptions)
-    newJobDescription
-  }
-
-  override def startJob(jobId: String): Unit = {
-    jobDescriptions = jobDescriptions.map { case (engineId, jds) =>
-      jds.get(jobId).fold(engineId -> jds) { d =>
-        engineId -> (jds + (jobId -> (d._1 -> d._2.copy(status = JobStatus.executing))))
-      }
-    }
-  }
+  private val jobDescriptions: TrieMap[String, Map[String, (Cancellable, JobDescription)]] = TrieMap.empty
 
-  override def startNewJob(engineId: String, f: Future[_], c: Cancellable, comment: String): JobDescription = {
-    val description = JobDescription(createUUID, JobStatus.executing, comment)
+  override def addJob(engineId: String, f: Future[_], c: Cancellable, comment: String): JobDescription = {
+    val description = JobDescription(createUUID, JobStatuses.executing, comment)
     val newJobDescriptions = jobDescriptions.getOrElse(engineId, Map.empty) +
-      (description.jobId -> (c -> description.copy(status = JobStatus.executing)))
-    jobDescriptions = jobDescriptions + (engineId -> newJobDescriptions)
+      (description.jobId -> (c -> description.copy(status = JobStatuses.executing)))
+    jobDescriptions.put(engineId, newJobDescriptions)
     f.onComplete {
       case Success(_) =>
         logger.info(s"Job ${description.jobId} completed successfully [engine $engineId]")
         removeJob(description.jobId)
       case Failure(e) =>
         logger.error(s"Job $description failed [engine $engineId]", e)
-        removeJob(description.jobId)
+        cancelJob(description.jobId)
     }
     description
   }
 
-  private def createUUID: String = UUID.randomUUID().toString
-
   /** Gets any active Jobs for the specified Engine */
   override def getActiveJobDescriptions(engineId: String): Map[String, JobDescription] = {
-    jobDescriptions.getOrElse(engineId, Map.empty).map {
-      case Tuple2(id, Tuple2(c, d)) => id -> d
-    }
+    (jobDescriptions.getOrElse(engineId, Map.empty).map {
+      case Tuple2(id, Tuple2(_, d)) => d
+    } ++ LivyJobServerSupport.status(engineId)).map(d => d.jobId -> d).toMap
   }
 
   override def removeJob(harnessJobId: String): Unit = {
-    jobDescriptions = jobDescriptions.map { case (engineId, jds) =>
+    jobDescriptions.map { case (engineId, jds) =>
       engineId -> (jds - harnessJobId)
     }
   }
 
   override def cancelJob(jobId: String): Future[Unit] = {
-    jobDescriptions.foreach {
+    Future.sequence(jobDescriptions.flatMap {
       case (_, jds) if jds.contains(jobId) =>
-        jds.get(jobId).foreach {
+        jds.get(jobId).map {
           case (cancellable, _) =>
             removeJob(jobId)
             cancellable.cancel()
         }
-    }
-    Future.successful(())
+    } ++ Seq(LivyJobServerSupport.cancel(jobId)))
+      .map(_ => ())
+      .recover {
+        case e: Exception =>
```

Comment:
@andreyshikov I researched this and

- we should use the monad `Try` where possible because both options for completion do not catch fatal errors. Not applicable here
- in the case above we should use `NonFatal` NOT `Exception` here since some fatal errors are caught by `Exception`

Reply:
Try extracts NonFatal..., ok let's use NonFatal (at least it's a common Scala practice)